### ISSUE TITLE
fix ordering of css dependencies

### ISSUE THREAD

**Issue by ianstormtaylor** [componentjs/builder.js#109]
right now, the ignoring that the builder does, while preventing dupes, doesn't result in the proper css output. for example, imagine a situation with a parent and two children...

`theme/component.json`:

``` json
{
  "name": "theme",
  "local": [
    "one",
    "two"
  ]
}
```

`theme/one/component.json`:

``` json
{
  "name": "one",
  "local": [
    "two"
  ],
  "styles": [
    "index.css"
  ]
}
```

`theme/one/index.css`:

``` css
.one {}
```

`theme/two/component.json`:

``` json
{
  "name": "two",
  "styles": [
    "index.css"
  ]
}
```

`theme/two/index.css`:

``` css
.two {}
```

when resolving the dependencies, the builder should run into `theme`'s two dependencies, and start with `one`. while evaluating `one` it should see that it depends on `two` and evaluate `two` first. the final output in `build/build.css` should be:

``` css
.two {}
.one {}
```

since `one` depends on `two`. instead, because `two` [gets ignored](https://github.com/component/builder.js/blob/master/lib/builder.js#L625) when evaluating `theme`, the actual output is:

``` css
.one {}
.two {}
```

and then styles in `one` that _should_ have overridden styles in `two` actually need `!important` flags to work. this makes it impossible (or annoying) to build up css dependencies, even though it would be a badass and obvious way to structure the css deps in themes

**Comment by tj** [comment]
hmm yeah that's definitely not ideal

**Comment by ianstormtaylor** [comment]
running into this more and more as i build out the segment.io theme - but i have no idea how to wrap my head around the recursion to fix it haha, but it will be badass once this works

**Comment by mnichols** [comment]
This just bit me on a large project too. Before manually stitching files to workaround I am going to have a look to see if I can patch

**Comment by jonathanong** [comment]
why even bother referencing locals in non-boot components when they're just CSS? just order them correctly in your `boot` component's `local` and you'll be fine.

**Comment by tj** [comment]
@jonathanong it's still pretty critical for some unfortunately since css is kinda lame haha, even simple stuff like normalization libs typically need to be first, or if you want to override something in a dependency then it has to be after. YAY CSS

**Comment by ianstormtaylor** [comment]
CSS, oh baby you!!
You don't got what I need
But you say your just a language
But you say your just a language
Oh CSS!!!

**Comment by tj** [comment]
hahaha nice song ian

**Comment by jonathanong** [comment]
gah now i'm getting this issue. time to spend time fixing component!

**Comment by micky2be** [comment]
Never run into this issue yet, but I barely override styles

**Comment by ianstormtaylor** [comment]
yeah this one is the one limiting issue for handling all the native types properly. after this and versioning are handled i'm pretty sure i have no "big picture" issues with component and everything is gravy after that

**Comment by ghost** [comment]
that would be really nice. i am building a project right now in which i have currently about 20-25 modules in my lib folder. almost all of them have some kind of styling. styles that should always appear "on the top" of the css, i simply put inside a `./lib/layout/[layout.css]` which is the first dependency of `./lib/boot` - BEM helps a lot here because the order (of the modules) in the compiled css is not that crucial. so it's not only a component(1) problem.

**Comment by necolas** [comment]
Is this still a problem in 0.12?

**Comment by peteschaffner** [comment]
@necolas yup still a problem in 0.12 -_-

**Comment by peteschaffner** [comment]
@jonathanong's [`flatten.js`](https://github.com/component/flatten.js) gets this right btw
